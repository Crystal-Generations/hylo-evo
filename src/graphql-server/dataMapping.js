import { invert } from 'lodash/fp'
import { currentDateString } from 'util/holochain'

// Data mapping from zome <> Hylo UI
export const zomeDefaultAttribsMap = () => ({
  post: {
    announcement: false,
    timestamp: currentDateString()
  },
  comment: {
    attachments: []
  }
})

export const toZomeKeyMap = {
  global: {
    'createdAt': 'timestamp',
    'id': 'address',
    'avatarUrl': 'avatar_url'
  },
  post: {
    'type': 'post_type'
  },
  person: {
    'id': 'agent_id'
  },
  comment: {
    'postId': 'base',
    'creator': 'agent_id'
  }
}

export const toUiKeyMap = {}
for (let key in toZomeKeyMap) {
  toUiKeyMap[key] = invert(toZomeKeyMap[key])
}

export const createDataRemapper = initialDataMap => (type, data) => {
  const dataMapForType = type in initialDataMap
    ? initialDataMap[type]
    : {}
  const dataMap = Object.assign({}, initialDataMap['global'], dataMapForType)
  let remappedData = {}

  for (let key in data) {
    const mappedKey = key in dataMap
      ? dataMap[key]
      : key
    remappedData[mappedKey] = data[key]
    // Mutating passed object. Used a cloned copy or do another way
    delete data[key]
  }

  return {
    ...data,
    ...remappedData
  }
}

export const toZomeData = (...args) => {
  const results = createDataRemapper(toZomeKeyMap)(...args)

  return {
    ...zomeDefaultAttribsMap()[args[0]],
    ...results
  }
}

export const toUiData = createDataRemapper(toUiKeyMap)

export const toUiQuerySet = results => ({
  total: results.length,
  items: results,
  hasMore: false
})
