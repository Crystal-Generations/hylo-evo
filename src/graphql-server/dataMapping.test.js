import { toUiKeyMap, toUiQuerySet, toZomeData, toUiData } from './dataMapping'

const uiData = {
  id: 'idhash',
  avatarUrl: 'avatarUrlthing'
}

const zomeData = {
  agent_id: 'idhash',
  avatar_url: 'avatarUrlthing'
}

const resultDataset = [
  { 1: 'test' },
  { 2: 'test' }
]

it('toZomeData should match snapshot', () => {

  expect(toZomeData('person', uiData)).toMatchSnapshot()
})

it('toUiData should match snapshot', () => {
  expect(toUiData('person', zomeData)).toMatchSnapshot()
})

it('toUiQuerySet should match snapshot', () => {
  expect(toUiQuerySet(resultDataset)).toMatchSnapshot()
})

it('toUiKeyMap should match snapshot', () => {

  expect(toUiKeyMap).toMatchSnapshot()
})
