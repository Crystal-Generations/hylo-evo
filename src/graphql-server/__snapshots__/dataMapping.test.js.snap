// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toUiData should match snapshot 1`] = `
Object {
  "avatarUrl": "avatarUrlthing",
  "id": "idhash",
}
`;

exports[`toUiKeyMap should match snapshot 1`] = `
Object {
  "comment": Object {
    "agent_id": "creator",
    "base": "postId",
  },
  "global": Object {
    "address": "id",
    "avatar_url": "avatarUrl",
    "timestamp": "createdAt",
  },
  "person": Object {
    "agent_id": "id",
  },
  "post": Object {
    "post_type": "type",
  },
}
`;

exports[`toUiQuerySet should match snapshot 1`] = `
Object {
  "hasMore": false,
  "items": Array [
    Object {
      "1": "test",
    },
    Object {
      "2": "test",
    },
  ],
  "total": 2,
}
`;

exports[`toZomeData should match snapshot 1`] = `
Object {
  "agent_id": "idhash",
  "avatar_url": "avatarUrlthing",
}
`;
